man/man5/rlm_files.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_files 5 "5 February 2004" "" "FreeRADIUS Module"
.SH NAME
rlm_files \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_files\fP module uses the 'users' file for accessing
authorization information for users.  Additionally, it supports
a 'users' file syntax to be applied to the accounting and pre-proxy
sections.
.PP
The main configuration items to be aware of are:
.IP usersfile
The filename of the 'users' file, which is parsed during the
authorization stage of this module.
.IP acctusersfile
The filename of the 'users' file, which is parsed during the
accounting stage of this module.
.IP preproxy_usersfile
The filename of the 'users' file, which is parsed during the
pre_proxy stage of this module.
.IP compat
This option allows FreeRADIUS to parse an old style Cistron syntax.
The default is 'no'.  If you need to parse an old style Cistron
file, set this option to 'cistron'.
.IP key
This option lets you set the attribute to use as a key to find
entries.  The default is "%{%{Stripped-User-Name}:-%{User-Name}}".  Note
that the key MUST supply real data.  Dynamic attributes like "Group"
will not work, because the "Group" attribute can only be used as a
comparison, to see if a user is in a Unix group.  It will not return
the name of the Unix group that a user is in.
.PP
If you want to use groups as a key, see the \fIrlm_passwd\fP, which
will create a real attribute that contains the group name.
.PP
This configuration entry enables you to have configurations that
perform per-group checks, and return per-group attributes, where the
group membership is dynamically defined by a previous module.  It also
lets you do things like key off of attributes in the reply, and
express policies like "when I send replies containing attribute
FOO with value BAR, do more checks, and maybe send additional
attributes".
.SH CONFIGURATION
.PP
.DS
modules {
  ... stuff here ...
.br
  files {
.br
    usersfile = %{confdir}/users
.br
    acctusersfile = %{confdir}/acct_users
.br
    preproxy_usersfile = %{confdir}/preproxy_users
.br
    compat = no
.br
    key = %{%{Stripped-User-Name}:-%{User-Name}}
.br
  }
.br
  ... stuff here ...
.br
}
.DE
.PP
.SH SECTIONS
.BR authorization,
.BR accounting,
.BR pre_proxy
.PP
.SH FILES
.I /etc/raddb/radiusd.conf,
.I /etc/raddb/users,
.I /etc/raddb/acct_users,
.I /etc/raddb/preproxy_users
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5),
.BR users (5)
.SH AUTHORS
Chris Parker, cparker@segv.org


doc/ChangeLog
FreeRADIUS 3.0.18 Tue 17 Apr 2018 14:00:00 EDT urgency=low
	Feature improvements
	* cleanup_delay can now be 30 seconds.  This helps
	  with proxies that have packet loss.
	* Do-Not-Respond policies can now be set in the
	  "post-auth" section.
	* Encode / Decode ADSL Forum DHCP options.
	* Fix module ordering issues.  e.g. when "sqlippool" needs
	  "sql".  See the "instantiate" section of radiusd.conf.
	* Add Big Switch dictionary.  Fixes #2252.
	* Add sql_session_start policy (raddb/policy.d/accounting)
	  This minimizes race conditions when using Simultaneous-Use
	  Patch from Philippe Wooding (#2257).
	* For rlm_perl, all variables are now tainted by default.  See
	  raddb/mods-available/perl, and the "perl_flags" configuration
	  item.  This change should only affect people who are using
	  variables in insecure ways.
	* Allow "sqlcounter" module to be listed in "post-auth".
	* Add support for IPv6 attributes in SQL.  Fixes #2280.
	  Patches from Michael Ducharme.
	* The server is better at handling fail-over for outbound
	  RadSec and TCP connections.  Fixes #2284.
	* The server is now more aggressive about retrying failed
	  outbound RadSec and TCP connections.  Fixes #2284.
	* Add TLS-Session-Version and TLS-Session-Cipher-Suite to
	  the "session_state" list.
	* Add expansion for Radsec connections.  "%{listen:TLS-...}"
	  for TLS-Client-Cert-* and TLS-Cert-* attributes.
	* Add notes on running "ldapsearch" using the parameters from
	  the LDAP module.
	* "ipaddr" attributes can now be cast to "integer" type attributes
	  in an "update" section.
	* Move main thread queue to using atomic queues.  This should help
	  with contention in high load scenarios.
	* Add "recv_buff" setting to listeners. For more details,
	  see sites-available/default
	* The sqlippool module can now use attributes other than "Pool-Name"
	  to assign IP pools.  The "Pool-Name" attribute is still the default.
	* The "unpack" expansion can now unpack substrings.  See
	  mods-available/unpack for documentation and examples.
	* The preprocess module now does "ciscvo_vsa_hack" for Eltex-AVPair.
	  Fixes #2301.  Vendors SHOULD NOT USE THAT KIND OF ATTRIBUTE.
	* Allow for <instance>-LDAP-UserDN.  See mods-available/ldap for
	  more information.

	Bug fixes
	* The session-state list is no longer cleaned in the
	  inner-tunnel.  This lets the outer Access-Reject
	  section access session-state.
	* Fix typo in lock initialization for TLS sockets
	  Found by Sergio NNX.
	* Add check for crash when home server down.
	  Fixes #2233.
	* Add username key for postauth table.
	* Better libpcap checks, when the header files or
	  libraries are missing.  Fixes #2245.
	* Allow building with old versions of OpenSSL.
	  Fixes #2247.
	* Allow non-FreeRADIUS State attributes to be used with
	  the "session-state" list.  i.e. State length != 16.
	* Be more aggressive about cleaning up zombie children
	  when running in debug mode.
	* Use LTDL_DEEPBIND, which fixes issues with Oracle
	  libraries exporting LDAP API functions.
	* unlock files when asked to unlock them.
	* return error instead of asserting in map code.
	* Don't write 0 bytes to SSL.  Fixes #2270.
	* Remove "expiry_time IS NULL" from allocate_update
	  query.  Fixes #2262.
	* Various dictionary cleanups and consistency checks.
	  Fixes #2281.  Patches from Peter Lemenkov.
	* rlm_python has stronger thread locking to prevent
	  reported issues.  Performance may be affected.
	* Don't allow Message-Authenticator to overflow
	  past the end of a large packet.
	* Fix crash in sqlippool when SQL server goes away.
	  Fixes #2300.
	* Typos in man pages.  Patch from Nikolai Kondrashov.
	  Fixes #2303.
	* Check for correct OpenSSL version in vulnerability
	  list.  Patch from Christian Hesse.
	* Fix crash with CoA packets/  Fixes #2304.

FreeRADIUS 3.0.17 Tue 17 Apr 2018 14:00:00 EDT urgency=low
	Feature improvements
	* Add CURLOPT_CAINFO.  Patch from Nicolas C.
	  #2167
	* "stats home server" now supports "src IPADDR",
	  to specify home server also by source IP.  Fixes #2169.
	* Add Dockerfiles for a selection of common systems.
	* Increase number of permitted file descriptors, for
	  systems with many home servers.
	* Add TLS-Client-Cert-X509v3-Extended-Key-Usage-OIDs.
	  Patch from Isaac Boukris.  Fixes #2205.
	* Update main READMEs.  Patches from Matthew Newton.
	* Added dictionary.mimosa
	* TLS-based EAP methods now create TLS-Session-Version
	  and TLS-Session-Cipher-Suite attributes.

	Bug fixes
	* Don't call post-proxy twice when proxying to
	  a virtual server.  Matthew Newton, #2161.
	* Use "raw" string value for shared secrets and dynamic clients.
	  It now parses strings with backslashes and "special characters"
	  correctly.  Fixes #2168.
	* Fix RuntimeDirectory for RedHat, from Alan Buxey.
	* Relax checks in 'if' parser from Isaac Bourkis
	* Minor cleanups for %{debug_attr:&request} from Isaac Boukris.
	* Be more aggressive about cleaning up cached certificate attributes,
	  due to deficiencies in OpenSSL.  Reported by Nicolas Reich.
	* Be more accepting when parsing IPv6 addresses.  Bug noted
	  by Klara Mall.
	* Fix double free in rlm_sql.  Fixes #2180.
	* rlm_detail now writes empty Access-Accept packets.
	* rlm_python can now create tagged attributes.
	* Don't crash on duplicate realm + authhost / accthost.
	  Bug found by Richard Palmer.
	* Allow partial certificate chain to trusted CA.  Fixes #2162
	* Treat SSL_read() returning zero as error.  Fixes #2164.
	* detail writer now checks if the file was renamed or deleted.
	* Add User-Name to Access-Accept if EAP-Message exists,
	  not Stripped-User-Name.
	* RedHat Systemd updates.  Fixes #2184
	* Use correct API for State variable in rlm_securid.
	* Remove broken radclient option "-i".
	* Fix "users" file (and hints, etc). So that it does not
	  get confused about entry ordering with multiple $INCLUDEs.
	* Fix rlm_sql to expand the un-escaped string, not the raw string.

src/modules/rlm_files/rlm_files.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or (at
 *   your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_files.c
 * @brief Process simple 'users' policy files.
 *
 * @copyright 2000,2006  The FreeRADIUS server project
 * @copyright 2000  Jeff Carneal <jeff@apex.net>
 */
RCSID("$Id$")

#include	<freeradius-devel/server/base.h>
#include	<freeradius-devel/server/modules.h>

#include	<ctype.h>
#include	<fcntl.h>

typedef struct rlm_files_t {
	char const *key;

	char const *filename;
	rbtree_t *common;

	/* autz */
	char const *usersfile;
	rbtree_t *users;


	/* authenticate */
	char const *auth_usersfile;
	rbtree_t *auth_users;

	/* preacct */
	char const *acct_usersfile;
	rbtree_t *acct_users;

#ifdef WITH_PROXY
	/* pre-proxy */
	char const *preproxy_usersfile;
	rbtree_t *preproxy_users;

	/* post-proxy */
	char const *postproxy_usersfile;
	rbtree_t *postproxy_users;
#endif

	/* post-authenticate */
	char const *postauth_usersfile;
	rbtree_t *postauth_users;
} rlm_files_t;

static fr_dict_t *dict_freeradius;

extern fr_dict_autoload_t rlm_files_dict[];
fr_dict_autoload_t rlm_files_dict[] = {
	{ .out = &dict_freeradius, .proto = "freeradius" },
	{ NULL }
};

static fr_dict_attr_t const *attr_fall_through;

extern fr_dict_attr_autoload_t rlm_files_dict_attr[];
fr_dict_attr_autoload_t rlm_files_dict_attr[] = {
	{ .out = &attr_fall_through, .name = "Fall-Through", .type = FR_TYPE_BOOL, .dict = &dict_freeradius },

	{ NULL }
};

/*
 *     See if a VALUE_PAIR list contains Fall-Through = Yes
 */
static int fall_through(VALUE_PAIR *vp)
{
	VALUE_PAIR *tmp;
	tmp = fr_pair_find_by_da(vp, attr_fall_through, TAG_ANY);

	return tmp ? tmp->vp_uint32 : 0;
}

static const CONF_PARSER module_config[] = {
	{ FR_CONF_OFFSET("filename", FR_TYPE_FILE_INPUT, rlm_files_t, filename) },
	{ FR_CONF_OFFSET("usersfile", FR_TYPE_FILE_INPUT, rlm_files_t, usersfile) },
	{ FR_CONF_OFFSET("acctusersfile", FR_TYPE_FILE_INPUT, rlm_files_t, acct_usersfile) },
#ifdef WITH_PROXY
	{ FR_CONF_OFFSET("preproxy_usersfile", FR_TYPE_FILE_INPUT, rlm_files_t, preproxy_usersfile) },
	{ FR_CONF_OFFSET("postproxy_usersfile", FR_TYPE_FILE_INPUT, rlm_files_t, postproxy_usersfile) },
#endif
	{ FR_CONF_OFFSET("auth_usersfile", FR_TYPE_FILE_INPUT, rlm_files_t, auth_usersfile) },
	{ FR_CONF_OFFSET("postauth_usersfile", FR_TYPE_FILE_INPUT, rlm_files_t, postauth_usersfile) },
	{ FR_CONF_OFFSET("key", FR_TYPE_STRING | FR_TYPE_XLAT, rlm_files_t, key) },
	CONF_PARSER_TERMINATOR
};


static int pairlist_cmp(void const *a, void const *b)
{
	return strcmp(((PAIR_LIST const *)a)->name, ((PAIR_LIST const *)b)->name);
}

static int getusersfile(TALLOC_CTX *ctx, char const *filename, rbtree_t **ptree)
{
	int rcode;
	PAIR_LIST *users = NULL;
	PAIR_LIST *entry, *next;
	PAIR_LIST *user_list, *default_list, **default_tail;
	rbtree_t *tree;

	if (!filename) {
		*ptree = NULL;
		return 0;
	}

	rcode = pairlist_read(ctx, filename, &users, 1);
	if (rcode < 0) {
		return -1;
	}

	/*
	 *	Walk through the 'users' file list, if we're debugging.
	 */
	if (rad_debug_lvl) {
		VALUE_PAIR *vp;

		entry = users;
		while (entry) {
			fr_cursor_t cursor;

			/*
			 *	Look for improper use of '=' in the
			 *	check items.  They should be using
			 *	'==' for on-the-wire RADIUS attributes,
			 *	and probably ':=' for server
			 *	configuration items.
			 */
			for (vp = fr_cursor_init(&cursor, &entry->check);
			     vp;
			     vp = fr_cursor_next(&cursor)) {
				/*
				 *	Ignore attributes which are set
				 *	properly.
				 */
				if (vp->op != T_OP_EQ) {
					continue;
				}

				/*
				 *	If it's a vendor attribute,
				 *	or it's a wire protocol,
				 *	ensure it has '=='.
				 */
				if ((fr_dict_vendor_num_by_da(vp->da) != 0) ||
				    (vp->da->attr < 0x100)) {
					WARN("[%s]:%d Changing '%s =' to '%s =='\n\tfor comparing RADIUS attribute in check item list for user %s",
					     filename, entry->lineno,
					     vp->da->name, vp->da->name,
					     entry->name);
					vp->op = T_OP_CMP_EQ;
					continue;
				}
			} /* end of loop over check items */

			/*
			 *	Look for server configuration items
			 *	in the reply list.
			 *
			 *	It's a common enough mistake, that it's
			 *	worth doing.
			 */
			for (vp = fr_cursor_init(&cursor, &entry->reply);
			     vp;
			     vp = fr_cursor_next(&cursor)) {
				/*
				 *	If it's NOT a vendor attribute,
				 *	and it's NOT a wire protocol
				 *	and we ignore Fall-Through,
				 *	then bitch about it, giving a
				 *	good warning message.
				 */
				 if (fr_dict_attr_is_top_level(vp->da) && (vp->da->attr > 1000)) {
					WARN("[%s]:%d Check item \"%s\"\n"
					       "\tfound in reply item list for user \"%s\".\n"
					       "\tThis attribute MUST go on the first line"
					       " with the other check items", filename, entry->lineno, vp->da->name,
					       entry->name);
				}
			}

			entry = entry->next;
		}
	}

	tree = rbtree_create(ctx, pairlist_cmp, NULL, RBTREE_FLAG_NONE);
	if (!tree) {
		pairlist_free(&users);
		return -1;
	}

	default_list = NULL;
	default_tail = &default_list;

	/*
	 *	We've read the entries in linearly, but putting them
	 *	into an indexed data structure would be much faster.
	 *	Let's go fix that now.
	 */
	for (entry = users; entry != NULL; entry = next) {
		/*
		 *	Remove this entry from the input list.
		 */
		next = entry->next;
		entry->next = NULL;
		(void) talloc_steal(tree, entry);

		/*
		 *	DEFAULT entries get their own list.
		 */
		if (strcmp(entry->name, "DEFAULT") == 0) {
			if (!default_list) {
				default_list = entry;

				/*
				 *	Insert the first DEFAULT into the tree.
				 */
				if (!rbtree_insert(tree, entry)) {
				error:
					pairlist_free(&entry);
					pairlist_free(&next);
					talloc_free(tree);
					return -1;
				}

			} else {
				/*
				 *	Tack this entry onto the tail
				 *	of the DEFAULT list.
				 */
				*default_tail = entry;
			}

			default_tail = &entry->next;
			continue;
		}

		/*
		 *	Not DEFAULT, must be a normal user.
		 */
		user_list = rbtree_finddata(tree, entry);
		if (!user_list) {
			/*
			 *	Insert the first one.
			 */
			if (!rbtree_insert(tree, entry)) goto error;
		} else {
			/*
			 *	Find the tail of this list, and add it
			 *	there.
			 */
			while (user_list->next) user_list = user_list->next;

			user_list->next = entry;
		}
	}

	*ptree = tree;

	return 0;
}



/*
 *	(Re-)read the "users" file into memory.
 */
static int mod_instantiate(void *instance, UNUSED CONF_SECTION *conf)
{
	rlm_files_t *inst = instance;

#undef READFILE
#define READFILE(_x, _y) do { if (getusersfile(inst, inst->_x, &inst->_y) != 0) { ERROR("Failed reading %s", inst->_x); return -1;} } while (0)

	READFILE(filename, common);
	READFILE(usersfile, users);
	READFILE(acct_usersfile, acct_users);

#ifdef WITH_PROXY
	READFILE(preproxy_usersfile, preproxy_users);
	READFILE(postproxy_usersfile, postproxy_users);
#endif

	READFILE(auth_usersfile, auth_users);
	READFILE(postauth_usersfile, postauth_users);

	return 0;
}

/*
 *	Common code called by everything below.
 */
static rlm_rcode_t file_common(rlm_files_t const *inst, REQUEST *request, char const *filename, rbtree_t *tree,
			       RADIUS_PACKET *request_packet, RADIUS_PACKET *reply_packet)
{
	char const	*name;
	VALUE_PAIR	*check_tmp = NULL;
	VALUE_PAIR	*reply_tmp = NULL;
	PAIR_LIST const *user_pl, *default_pl;
	bool		found = false;
	PAIR_LIST	my_pl;
	char		buffer[256];

	if (!inst->key) {
		VALUE_PAIR	*namepair;

		namepair = request->username;
		name = namepair ? namepair->vp_strvalue : "NONE";
	} else {
		int len;

		len = xlat_eval(buffer, sizeof(buffer), request, inst->key, NULL, NULL);
		if (len < 0) {
			return RLM_MODULE_FAIL;
		}

		name = len ? buffer : "NONE";
	}

	if (!tree) return RLM_MODULE_NOOP;

	my_pl.name = name;
	user_pl = rbtree_finddata(tree, &my_pl);
	my_pl.name = "DEFAULT";
	default_pl = rbtree_finddata(tree, &my_pl);

	/*
	 *	Find the entry for the user.
	 */
	while (user_pl || default_pl) {
		fr_cursor_t cursor;
		VALUE_PAIR *vp;
		PAIR_LIST const *pl;

		/*
		 *	Figure out which entry to match on.
		 */

		if (!default_pl && user_pl) {
			pl = user_pl;
			user_pl = user_pl->next;

		} else if (!user_pl && default_pl) {
			pl = default_pl;
			default_pl = default_pl->next;

		} else if (user_pl->order < default_pl->order) {
			pl = user_pl;
			user_pl = user_pl->next;

		} else {
			pl = default_pl;
			default_pl = default_pl->next;
		}

		MEM(fr_pair_list_copy(request, &check_tmp, pl->check) >= 0);
		for (vp = fr_cursor_init(&cursor, &check_tmp);
		     vp;
		     vp = fr_cursor_next(&cursor)) {
			if (xlat_eval_pair(request, vp) < 0) {
				RWARN("Failed parsing expanded value for check item, skipping entry: %s", fr_strerror());
				fr_pair_list_free(&check_tmp);
				continue;
			}
		}

		if (paircmp(request, request_packet->vps, check_tmp, &reply_packet->vps) == 0) {
			RDEBUG2("Found match \"%s\" one line %d of %s", pl->name, pl->lineno, filename);
			found = true;

			/* ctx may be reply or proxy */
			MEM(fr_pair_list_copy(reply_packet, &reply_tmp, pl->reply) >= 0);

			radius_pairmove(request, &reply_packet->vps, reply_tmp, true);
			fr_pair_list_move(request, &request->control, &check_tmp);

			reply_tmp = NULL;	/* radius_pairmove() frees input attributes */
			fr_pair_list_free(&check_tmp);

			/*
			 *	Fallthrough?
			 */
			if (!fall_through(pl->reply)) break;
		}
	}

	/*
	 *	Remove server internal parameters.
	 */
	fr_pair_delete_by_da(&reply_packet->vps, attr_fall_through);

	/*
	 *	See if we succeeded.
	 */
	if (!found)
		return RLM_MODULE_NOOP; /* on to the next module */

	return RLM_MODULE_OK;

}


/*
 *	Find the named user in the database.  Create the
 *	set of attribute-value pairs to check and reply with
 *	for this user from the database. The main code only
 *	needs to check the password, the rest is done here.
 */
static rlm_rcode_t CC_HINT(nonnull) mod_authorize(void *instance, UNUSED void *thread, REQUEST *request)
{
	rlm_files_t const *inst = instance;

	return file_common(inst, request, inst->filename,
			   inst->users ? inst->users : inst->common,
			   request->packet, request->reply);
}


/*
 *	Pre-Accounting - read the acct_users file for check_items and
 *	config. Reply items are Not Recommended(TM) in acct_users,
 *	except for Fallthrough, which should work
 */
static rlm_rcode_t CC_HINT(nonnull) mod_preacct(void *instance, UNUSED void *thread, REQUEST *request)
{
	rlm_files_t const *inst = instance;

	return file_common(inst, request, inst->acct_usersfile,
			   inst->acct_users ? inst->acct_users : inst->common,
			   request->packet, request->reply);
}

#ifdef WITH_PROXY
static rlm_rcode_t CC_HINT(nonnull) mod_pre_proxy(void *instance, UNUSED void *thread, REQUEST *request)
{
	rlm_files_t const *inst = instance;

	return file_common(inst, request, inst->preproxy_usersfile,
			   inst->preproxy_users ? inst->preproxy_users : inst->common,
			   request->packet, request->proxy->packet);
}

static rlm_rcode_t CC_HINT(nonnull) mod_post_proxy(void *instance, UNUSED void *thread, REQUEST *request)
{
	rlm_files_t const *inst = instance;

	return file_common(inst, request, inst->postproxy_usersfile,
			   inst->postproxy_users ? inst->postproxy_users : inst->common,
			   request->proxy->reply, request->reply);
}
#endif

static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(void *instance, UNUSED void *thread, REQUEST *request)
{
	rlm_files_t const *inst = instance;

	return file_common(inst, request, inst->auth_usersfile,
			   inst->auth_users ? inst->auth_users : inst->common,
			   request->packet, request->reply);
}

static rlm_rcode_t CC_HINT(nonnull) mod_post_auth(void *instance, UNUSED void *thread, REQUEST *request)
{
	rlm_files_t const *inst = instance;

	return file_common(inst, request, inst->postauth_usersfile,
			   inst->postauth_users ? inst->postauth_users : inst->common,
			   request->packet, request->reply);
}


/* globally exported name */
extern rad_module_t rlm_files;
rad_module_t rlm_files = {
	.magic		= RLM_MODULE_INIT,
	.name		= "files",
	.inst_size	= sizeof(rlm_files_t),
	.config		= module_config,
	.instantiate	= mod_instantiate,
	.methods = {
		[MOD_AUTHENTICATE]	= mod_authenticate,
		[MOD_AUTHORIZE]		= mod_authorize,
		[MOD_PREACCT]		= mod_preacct,

#ifdef WITH_PROXY
		[MOD_PRE_PROXY]		= mod_pre_proxy,
		[MOD_POST_PROXY]	= mod_post_proxy,
#endif
		[MOD_POST_AUTH]		= mod_post_auth
	},
};

